packages/react-components/react-list-preview/src/components/ListItem/useListItem.tsx

```diff
@@ -0,0 +1,218 @@
+import * as React from 'react';
+import {
+  dispatchGroupperMoveFocusEvent,
+  dispatchMoverMoveFocusEvent,
+  TabsterDOMAttribute,
+  TabsterTypes,
+  useArrowNavigationGroup,
+  useFocusableGroup,
+  useFocusFinders,
+  useMergedTabsterAttributes_unstable,
+} from '@fluentui/react-tabster';
+import {
+  getIntrinsicElementProps,
+  mergeCallbacks,
+  slot,
+  useEventCallback,
+  useId,
+  useMergedRefs,
+} from '@fluentui/react-utilities';
+import type { ListItemProps, ListItemState } from './ListItem.types';
+import { useListContext_unstable } from '../List/listContext';
+import { Enter, Space, ArrowUp, ArrowDown, ArrowRight, ArrowLeft } from '@fluentui/keyboard-keys';
+import { Checkbox, CheckboxOnChangeData } from '@fluentui/react-checkbox';
+
+const DEFAULT_ROOT_EL_TYPE = 'li';
+
+function validateProperElementTypes(parentRenderedAs?: 'div' | 'ul' | 'ol', renderedAs?: 'div' | 'li') {
+  if (process.env.NODE_ENV === 'production') {
+    return;
+  }
+
+  if (renderedAs === 'div' && parentRenderedAs !== 'div') {
+    throw new Error('ListItem cannot be rendered as a div when its parent is not a div.');
+  }
+  if (renderedAs === 'li' && parentRenderedAs === 'div') {
+    throw new Error('ListItem cannot be rendered as a li when its parent is a div.');
+  }
+}
+
+/**
+ * Create the state required to render ListItem.
+ *
+ * The returned state can be modified with hooks such as useListItemStyles_unstable,
+ * before being passed to renderListItem_unstable.
+ *
+ * @param props - props from this instance of ListItem
+ * @param ref - reference to root HTMLLIElement | HTMLDivElementof ListItem
+ */
+export const useListItem_unstable = (
+  props: ListItemProps,
+  ref: React.Ref<HTMLLIElement | HTMLDivElement>,
+): ListItemState => {
+  const id = useId('listItem');
+  const { value = id, onKeyDown, onClick, tabIndex } = props;
+
+  const toggleItem = useListContext_unstable(ctx => ctx.selection?.toggleItem);
+  const isSelectionEnabled = useListContext_unstable(ctx => !!ctx.selection);
+  const isSelected = useListContext_unstable(ctx => ctx.selection?.isSelected(value));
+  const listItemRole = useListContext_unstable(ctx => ctx.accessibilityRoles.listItemRole);
+  const setFocusableChildren = useListContext_unstable(ctx => ctx.accessibilityRoles.setFocusableChildren);
+
+  const { findAllFocusable } = useFocusFinders();
+
+  const focusableItems = isSelectionEnabled || tabIndex === 0 || onClick || onKeyDown;
+
+  const parentRenderedAs = useListContext_unstable(ctx => ctx.as);
+  const renderedAs = props.as || DEFAULT_ROOT_EL_TYPE;
+
+  const rootRef = React.useRef<HTMLLIElement | HTMLDivElement>(null);
+
+  validateProperElementTypes(parentRenderedAs, renderedAs);
+
+  React.useEffect(() => {
+    if (rootRef.current) {
+      const focusable = findAllFocusable(rootRef.current);
+      setFocusableChildren(focusable.length > 0);
```

Comment:
Let's use focusable detection in development mode only to throw. Otherwise we'll end up in a situation where every single list item will double render on mount

Reply:
that way we will not be able to infer the roles however, and the developers will always have to pass them. I guess that is acceptable if we want the best perf.

Reply:
Can the roles be toggled by user prop instead from the `List`?

A development throw should be sufficient at least to let folks know something is wrong and that they should set a prop

Reply:
sure, its possible. I was hoping for a seamless solution here where we would automatically use the proper role based on the use case, however I don't see how we can do that without rerenders.

I guess we could make the `role` prop required on `List`, this will also allow us to derive the roles on the list items themselves. Then after render and if in dev mode, I can warn the user if improper role seems to be used. This should be a warning however as we don't want to prevent people from using other roles if their specific use case calls for it.

Reply:
ended up going with the `navigationMode` prop, which dictates the default role for List and ListItem. `grid` whenever `navigationMode` is `"composite"` and `list` or `listbox` when it is `"items"` (based on if selection is enabled).